modules/ephemeris/src/main/scala/gem/horizons/HorizonsEphemerisQuery.scala

```diff
@@ -83,76 +117,100 @@ object HorizonsEphemerisQuery {
+    * @param step  time (ms precision) between successive ephemeris elements (a
+    *              minimum of 500ms will be used if a smaller time is requested)
     *
-    * @return an IO that computes an Horizons query ready to be executed
+    * @return List of queries that cover the entire time range
     */
-  def forCurrentSemesterWithPadding(key:   HorizonsDesignation,
-                                    site:  Site,
-                                    limit: Int): IO[HorizonsEphemerisQuery] =
-
-    Semester.current(site).map { s =>
-      apply(
-        key,
-        site,
-        s.start.localDateTime.minusMonths(1L).toInstant(UTC),
-        s.end.localDateTime.plusMonths(1L).toInstant(UTC),
-        limit
-      )
+  def paging(
+        key:   HorizonsDesignation,
+        site:  Site,
+        start: Instant,
+        end:   Instant,
+        step:  Duration): List[HorizonsEphemerisQuery] = {
+
+    // Horizons doesn't support a step length below 500 ms and FRACSEC gives
+    // only ms precision.
+    val stepʹ   = roundUpMs(step) max MinStepLen
+    val stepMsʹ = stepʹ.toMillis
+
+    // Total time covered by a single page with max elements.
+    val maxPage  = stepʹ * MaxElements.toLong
+
+    // Calculate (end time inclusive, element count) for a page starting at pStart.
+    def pageEnd(pStart: Instant): (Instant, Int) = {
+      val rem = roundUpMs(Duration.between(pStart, end))
+      val cnt = if (rem >= maxPage) MaxElements.toLong else intCeil(rem.toMillis, stepMsʹ) + 1L
+
+      (pStart + stepʹ * (cnt - 1), cnt.toInt)
     }
 
+    Stream.unfold(start) { pStart =>
+      if (pStart >= end)
+        None
+      else {
+        val (pEnd, pCnt) = pageEnd(pStart)
+
+        // Horizons queries must return >= 2 elements.  If next page would have
+        // just 1 elem, shorten *this* query so that the next one has 2.
+        val (_,     nCnt ) = pageEnd(pStart + maxPage)
+        val (pEndʹ, pCntʹ) = if (nCnt === 1) (pEnd - stepʹ, pCnt - 1) else (pEnd, pCnt)
+
+        Some((HorizonsEphemerisQuery(key, site, pStart, pEndʹ, pCntʹ), pEndʹ + stepʹ))
+      }
+    }.toList
+  }
+
+  def pagingSemester(
+        key:      HorizonsDesignation,
+        site:     Site,
+        semester: Semester,
+        step:     Duration,
+        padding:  Period): List[HorizonsEphemerisQuery] = {
+
+    val start = semester.start.atSite(site).minus(padding).toInstant
+    val end   = semester.end.atSite(site).plus(padding).toInstant
+
+    paging(key, site, start, end, step)
+  }
+
+  def pagingCurrentSemester(
+        key:      HorizonsDesignation,
+        site:     Site,
+        step:     Duration,
+        padding:  Period): IO[List[HorizonsEphemerisQuery]] =
+
+    Semester.current(site).map(pagingSemester(key, site, _, step, padding))
+
+  def streamAll(qs: List[HorizonsEphemerisQuery]): Stream[IO, Ephemeris.Element] =
+    Monoid.combineAll(qs.map(_.streamEphemeris))
```

Comment:
This should be equivalent to `qs.foldMap(_.streamEphemeris)`

Reply:
Right, thanks.  I'll update it in the next one.